### PR TITLE
Request for initial review: Bucket types

### PR DESCRIPTION
**Pull request by bookshelfdave**
Add a bucket type parameter to the following PB message:
- get
- put
- delete
- listkeys
- listbuckets
- 2i

Additionally, all references to 2index were changes to 2i. Perhaps this should have been done on a separate PR, but I needed up update 2i code anyways. Sorry for the noise this creates in the PR. 

I'd like to talk w/ @javajolt a bit more re: integration testing of each message, which I'll add in a later commit (hence the "request for initial review"). For now, existing integration tests take a `NULL` bucket type, indicating the `default` bucket.

### CODE REVIEW

File: examples/async_example.c [hunk 1]
```diff
@@ -79,6 +79,10 @@
 
     // create some sample binary values to use
     riak_binary *bucket_bin   = riak_binary_copy_from_string(cfg, args.bucket); // Not copied
+    riak_binary *bucket_type_bin = NULL;
+    if(args.has_bucket_type) {
+        bucket_type_bin = riak_binary_copy_from_string(cfg, args.bucket_type);
```
Comment (hazen) [inline examples/async_example.c:84]: Looks like if the `args.bucket_type` is NULL, then `bucket_type_bin` is automatically NULL.  This is how all of the other arguments work, so the if statement, while clearer, is superfluous.

File: examples/example.c [hunk 2]
```diff
@@ -61,6 +61,10 @@
 
     // create some sample binary values to use
     riak_binary *bucket_bin   = riak_binary_copy_from_string(cfg, args.bucket); // Not copied
+    riak_binary *bucket_type_bin = NULL;
+    if(args.has_bucket_type) {
```
Comment (hazen) [inline examples/example.c:65]: As in `async_example.c`, the if statement is not necessary

**Comment by hazen** [comment]
I asked in **Eng:Clients** and found that bucket types are higher order namespaces than buckets.  So I would order your arguments/storage: _bucket type_, _bucket_, _key_, i.e., type before bucket across the board

File: examples/async_example.c [hunk 3]
```diff
@@ -151,7 +155,11 @@
             }
             riak_get_options_set_basic_quorum(get_options, RIAK_TRUE);
             riak_get_options_set_r(get_options, 2);
-            err = riak_async_register_get(rop, bucket_bin, key_bin, get_options, (riak_response_callback)example_get_cb);
+            err = riak_async_register_get(rop,
+                                          bucket_bin,
+                                          bucket_type_bin,
+                                          key_bin, get_options,
```
Comment (hazen) [inline examples/async_example.c:161]: 1 per line?

**Comment by hazen** [comment]
We'll probably want to have integration tests with bucket types and without.  Since we cannot setup the bucket type within the client, we still need to figure out the test setup to make that happen.

**Comment by hazen** [comment]
Weirdly the integration test cored on 2.0.0.pre11, but work fine on 2.0.0pre18.

**Comment by hazen** [comment]
@metadave - How about we merge this to get the 2i naming changes in and the initial bucket type changes in?  We can reorder the parameters in the next PR and add more tests.

**Comment by bookshelfdave** [comment]
@javajolt `test_integration_async_listkeys` is coredumping somehow, it looks like `riak_listkeys_request_encode` is getting a non-NULL bucket_type

**Comment by hazen** [comment]
Which Riak version are you testing against?

**Comment by bookshelfdave** [comment]
`develop`, but I think the bug is somewhere in the test code.

**Comment by hazen** [comment]
Agreed.  Should not core.  However, I ran the tests and got the segfault, now the 2.0.0pre18 node I was talking to became unresponsive:  

```
curl http://127.0.0.1:10018/buckets?buckets=true
curl: (7) couldn't connect to host
```

**Pull request closed without merging**